### PR TITLE
BlobRoute : Comparison and Hash can be made much cheaper without string formatting.

### PR DESCRIPTION
**Pull request by szmyd**
@sanebay we can reuse this class and derive from BtreeKey in HomeStore (or that's the goal).

### CODE REVIEW

File: src/lib/memory/homeobject.hpp [hunk 1]
```diff
@@ -91,3 +67,25 @@ class MemoryHomeObject : public HomeObjectImpl {
...
+    auto format(const homeobject::BlobRoute& r, FormatContext& ctx) {
+        return format_to(ctx.out(), "{}:{:012x}", r.shard, r.blob);
+    }
+};
+} // namespace fmt
+
+template <>
+struct std::hash< homeobject::BlobRoute > {
+    std::size_t operator()(homeobject::BlobRoute const& r) const noexcept {
+        return std::hash< homeobject::shard_id >{}(r.shard) ^ (std::hash< homeobject::blob_id >{}(r.blob) << 1);
```
Comment (szmyd) [inline src/lib/memory/homeobject.hpp:89]: This arithmetic lifted from: https://en.cppreference.com/w/cpp/utility/hash

**Comment by szmyd** [comment]
Example logs:
```
[09/20/23 22:25:01.293] [bin/blob_memory_test] [trace] [67814] [blob_manager.cpp:17:_put_blob] Writing Blob 0001:000000000001:00000000000020a4
[09/20/23 22:25:01.291] [bin/blob_memory_test] [trace] [67773] [blob_manager.cpp:39:_get_blob] Looking up Blob 0001:000000000000:00000000000007fb
[09/20/23 22:25:01.295] [bin/blob_memory_test] [trace] [67806] [blob_manager.cpp:39:_get_blob] Looking up Blob 0001:000000000001:0000000000000842
[09/20/23 22:25:01.295] [bin/blob_memory_test] [warning] [67773] [blob_manager.cpp:42:_get_blob] Blob [0001:000000000000:00000000000007fb] missing during get
[09/20/23 22:25:01.294] [bin/blob_memory_test] [warning] [67753] [blob_manager.cpp:42:_get_blob] Blob [0001:000000000000:000000000000083a] missing during get
[09/20/23 22:25:01.293] [bin/blob_memory_test] [trace] [67760] [blob_manager.cpp:39:_get_blob] Looking up Blob 0001:000000000000:0000000000000000
[09/20/23 22:25:01.293] [bin/blob_memory_test] [trace] [67797] [blob_manager.cpp:61:_del_blob] Tombstoning Blob 0001:000000000000:000000000000083c
[09/20/23 22:25:01.295] [bin/blob_memory_test] [trace] [67790] [blob_manager.cpp:39:_get_blob] Looking up Blob 0001:000000000001:0000000000000805
[09/20/23 22:25:01.294] [bin/blob_memory_test] [warning] [67835] [blob_manager.cpp:68:_del_blob] Blob [0001:000000000000:0000000000000835] missing during del
[09/20/23 22:25:01.295] [bin/blob_memory_test] [trace] [67797] [blob_manager.cpp:61:_del_blob] Tombstoning Blob 0001:000000000000:0000000000000838
```

File: src/lib/homestore/homeobject.hpp [hunk 2]
```diff
@@ -27,7 +27,6 @@ class HSHomeObject : public HomeObjectImpl {
 
 private:
     shard_id generate_new_shard_id(pg_id pg);
-    shard_id make_new_shard_id(pg_id pg, uint64_t sequence_num) const;
     uint64_t get_sequence_num_from_shard_id(uint64_t shard_id);
```
Comment (xiaoxichen) [inline src/lib/homestore/homeobject.hpp:30]: s/uint64_t/shard_id?
maybe shard_id_t..

File: src/lib/blob_route.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                         (r.shard & homeobject::shard_mask), r.blob);
+    }
+};
+} // namespace fmt
+
+template <>
+struct std::hash< homeobject::BlobRoute > {
+    std::size_t operator()(homeobject::BlobRoute const& r) const noexcept {
+        // This arithmetic lifted from: https://en.cppreference.com/w/cpp/utility/hash
+        return std::hash< homeobject::shard_id >{}(r.shard) ^ (std::hash< homeobject::blob_id >{}(r.blob) << 1);
```
Comment (xiaoxichen) [inline src/lib/blob_route.hpp:38]: It really doesnt matter :) ,  below copying from boost::hash_combine
```
std::size_t operator()(homeobject::BlobRoute const& r) const noexcept {
    auto seed = std::hash< homeobject::shard_id >{}(r.shard)
    seed ^= std::hash< homeobject::blob_id >{}(r.blob) + 0x9e3779b9 + (seed<<6) + (seed>>2);
    return seed
}
```

**Review by xiaoxichen — APPROVED** [review]
LGTM

File: src/lib/blob_route.hpp (hunk 3 above)
Comment (szmyd) [reply src/lib/blob_route.hpp:38]: I didn't spend any time looking at an analysis, is boost::combine better? (Uniform Distribution)?

Comment (xiaoxichen) [reply src/lib/blob_route.hpp:38]: I was just look at boost::combine_hash when looking at Hari's MultiBlk implementation. 
Yes it is , sort of maths , the 0x9e3779b9 is  Golden Ratio's fractional part which can better distribute into int64 space.

**Pull request merged**